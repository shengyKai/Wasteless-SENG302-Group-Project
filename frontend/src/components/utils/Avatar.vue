<template>
  <v-avatar
    :size="sizeInPixels"
    :color="user !== undefined ? 'primary' : 'secondary'"
    class="white--text headline"
    :style="`font-size: ${fontSize}px !important`"
  >
    <template v-if="image === undefined">{{ initials }}</template>
    <v-img v-else :src="image" height="100%"/>
  </v-avatar>
</template>

<script>
import { imageSrcFromFilename } from '@/utils';

export default {
  name: 'Avatar',
  props: {
    /**
     * User to make avatar for.
     * Could be aquired from the 'getUser' function in api.ts
     */
    user: Object,

    /**
     * Business to make avatar for.
     */
    business: Object,


    /**
     * The size of the avatar
     * The small and medium sizes use the thumbnail sized avatar image, large uses full size.
     * @values small, medium, large
     */
    size: {
      validator: (s) => ['small', 'medium', 'large'].includes(s),
      default: 'medium',
    },
  },

  computed: {
    sizeInPixels() {
      switch (this.size) {
      case "small":
        return 30;
      case "medium":
        return 40;
      case "large":
        return 200;
      default:
        throw new Error(`Invalid size "${this.size}"`);
      }
    },
    fontSize() {
      return this.sizeInPixels * 0.5;
    },
    initials() {
      if (this.user !== undefined) {
        return this.user.firstName[0].toUpperCase() + this.user.lastName[0].toUpperCase();
      } else if (this.business !== undefined) {
        let pieces = this.business.name.split(/\s/);
        if (pieces.length === 2) {
          return pieces[0][0].toUpperCase() + pieces[1][0].toUpperCase();
        }
        return this.business.name[0].toUpperCase();
      } else {
        throw new Error('Either "user" or "business" must be provided.');
      }
    },
    image() {
      let image;
      if (this.user !== undefined) {
        image = this.user.images[0];
      } else if (this.business !== undefined) {
        image = this.business.images[0];
      } else {
        throw new Error('Either "user" or "business" must be provided.');
      }

      if (image === undefined) return undefined;

      let url = image.thumbnailFilename;
      if (url === undefined || url === null) {
        url = image.filename;
      }
      return imageSrcFromFilename(url);
    }
  }
};
</script>

<style scoped>
.profile-image {
  margin: auto;
  position: absolute;
  top: 0; left: 0; bottom: 0; right: 0;
}
</style>