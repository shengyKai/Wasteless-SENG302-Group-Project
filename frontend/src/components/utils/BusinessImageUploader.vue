<template>
  <v-dialog v-model="showDialog" persistent max-width="500px">
    <v-form>
      <v-card>
        <v-card-title>
          Upload Business Image
        </v-card-title>
        <v-card-text>
          <ImageSelector v-model="file"/>
        </v-card-text>
        <v-card-actions>
          <v-spacer/>
          <!-- Current error message -->
          <div class="error-text" v-if="errorMessage !== undefined"> {{ errorMessage }} </div>
          <v-spacer/>
          <v-btn
            color="primary"
            text
            @click="closeForm">
            Close
          </v-btn>
          <v-btn
            type="submit"
            color="primary"
            :disabled="!file"
            :loading="isLoading"
            @click.prevent="uploadImage">
            Upload
          </v-btn>
        </v-card-actions>
      </v-card>
    </v-form>
  </v-dialog>
</template>

<script>
import ImageSelector from "@/components/utils/ImageSelector";

export default {
  name: "BusinessImageUploader",
  components: {ImageSelector},
  props: {
    value: undefined,
  },
  data() {
    return {
      isLoading: false,
      errorMessage: undefined,
      showDialog: true
    };
  },
  methods: {
    /*
     * Close the dialog and emit message to the parent to add file to uploaded images.
     */
    uploadImage() {
      this.$emit('uploadImage');
    },
    /**
     * Closes the dialog and clears any selected files
     */
    closeForm() {
      this.file = undefined;
      this.$emit('closeDialog');
    },
  },
  computed: {
    file: {
      get() {
        return this.value;
      },
      set (value) {
        this.$emit('input', value);
      },
    }
  }
};
</script>