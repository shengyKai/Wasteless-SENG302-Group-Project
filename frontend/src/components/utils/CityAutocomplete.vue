<template>
  <v-combobox
    label="City"
    :items="cityItems"
    :loading="isLoading"
    :search-input.sync="citySearch"
    :rules="mandatoryRules.concat(maxCharRule)"
    no-filter
    clearable
    outlined
  />
</template>

<script>
export default {
  name: 'CityAutocomplete',
  data () {
    return {
      cityItems: [],
      isLoading: false,
      citySearch: null,
      mandatoryRules: [
        //All fields with the class "required" will go through this ruleset to ensure the field is not empty.
        //if it does not follow the format, display error message
        field => !!field || 'Field is required'
      ],
      maxCharRule: [
        field => (field.length <= 100) || 'Reached max character limit: 100'
      ]
    };
  },

  watch: {
    citySearch (val) {
      this.cityItems = [];

      //if input length exists, and has a length more than 2
      if (val && val.length > 2) {
        //show loading animation
        this.isLoading = true;
        //append to the api url the input the user has entered
        let url = `https://photon.komoot.io/api/?q=${encodeURIComponent(val)}&osm_tag=place:city&osm_tag=place:town`;

        fetch(url).then(res => res.json()).then(res => {
          res.features.forEach(feature => {
            //If the returned GEOJSON has any of these key(s):
            //city
            //it will add that result into the autocomplete suggestion
            this.cityItems.push(feature.properties.name);
          });
        }).catch(err => {
          console.log(err);
          //after everything is shown, the loading animation will stop
        }).finally(() => (this.isLoading = false));
      }
    },
  }
};
</script>
