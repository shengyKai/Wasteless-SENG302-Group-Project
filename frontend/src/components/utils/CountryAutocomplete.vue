<template>
  <v-combobox
    label="Country"
    :items="countryItems"
    :loading="isLoading"
    :search-input.sync="countrySearch"
    :rules="mandatoryRules.concat(maxCharRule)"
    no-filter
    clearable
    outlined
  />
</template>

<script>
export default {
  name: 'CountryAutocomplete',
  data () {
    return {
      countryItems: [],
      isLoading: false,
      countrySearch: null,
      mandatoryRules: [
        //All fields with the class "required" will go through this ruleset to ensure the field is not empty.
        //if it does not follow the format, display error message
        field => !!field || 'Field is required'
      ],
      maxCharRule: [
        field => (field.length <= 100) || 'Reached max character limit: 100'
      ]
    };
  },

  watch: {
    countrySearch (val) {
      this.countryItems = [];

      //if input length exists, and has a length more than 2
      if (val && val.length > 2) {
        //show loading animation
        this.isLoading = true;
        //append to the api url the input the user has entered
        let url = `https://photon.komoot.io/api/?q=${encodeURIComponent(val)}&osm_tag=place:country`;

        fetch(url).then(res => res.json()).then(res => {
          res.features.forEach(feature => {
            //If the returned GEOJSON has any of these key(s):
            //country
            //it will add that result into the autocomplete suggestion
            this.countryItems.push(feature.properties.name);
          });
        }).catch(err => {
          console.log(err);
          //after everything is shown, the loading animation will stop
        }).finally(() => (this.isLoading = false));
      }
    },
  }
};
</script>
