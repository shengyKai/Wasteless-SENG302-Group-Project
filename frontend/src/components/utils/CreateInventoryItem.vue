<template>
  <v-row justify="center">
    <v-dialog
      v-model="dialog"
      persistent
      max-width="600px"
    >
      <v-form v-model="valid">
        <v-card>
          <v-card-title>
            <h4 class="primary--text">Create Inventory Item</h4>
          </v-card-title>
          <v-card-text>
            <v-container>
              <v-row>
                <v-col>
                  <v-text-field
                    class="required"
                    v-model="productCode"
                    label="Product Code"
                    :rules="mandatoryRules.concat(maxCharRules)"
                    outlined
                  />
                </v-col>
                <v-col cols="12">
                  <v-textarea
                    v-model="quantity"
                    label="Quantity"
                    :rules="maxCharDescriptionRules"
                    rows="3"
                    outlined
                  />
                </v-col>
                <v-col cols="12">
                  <v-select
                    class="required"
                    v-model="pricePerItem"
                    :items="businessTypes"
                    label="Price per Item"
                    :rules="mandatoryRules.concat(maxCharRules)"
                    outlined
                  />
                </v-col>
                <v-col cols="12">
                  <v-text-field
                    class="required"
                    v-model="totalPrice"
                    label="Total Price"
                    :rules="mandatoryRules.concat(streetRules)"
                    outlined/>
                </v-col>
                <v-col cols="12">
                  <v-text-field
                    class="required"
                    v-model="manufactured"
                    label="Manufactured"
                    :rules="mandatoryRules.concat(streetRules)"
                    outlined/>
                </v-col>
                <v-col cols="12">
                  <v-text-field
                    class="required"
                    v-model="sellBy"
                    label="Sell By"
                    :rules="mandatoryRules.concat(streetRules)"
                    outlined/>
                </v-col>
                <v-col cols="12">
                  <v-text-field
                    class="required"
                    v-model="bestBefore"
                    label="Best Before"
                    :rules="mandatoryRules.concat(streetRules)"
                    outlined/>
                </v-col>
                <v-col cols="12">
                  <v-text-field
                    class="required"
                    v-model="expires"
                    label="Expires"
                    :rules="mandatoryRules.concat(streetRules)"
                    outlined/>
                </v-col>
              </v-row>
              <p class="error-text" v-if ="errorMessage !== undefined"> {{errorMessage}} </p>
            </v-container>
          </v-card-text>
          <v-card-actions>
            <v-spacer/>
            <v-btn
              color="primary"
              text
              @click="closeDialog">
              Close
            </v-btn>
            <v-btn
              type="submit"
              color="primary"
              :disabled="!valid"
              @click.prevent="createBusiness">
              Create
            </v-btn>
          </v-card-actions>
        </v-card>
      </v-form>
    </v-dialog>
  </v-row>
</template>

<script>
import {createBusiness} from '@/api/internal';

export default {
  name: 'CreateBusiness',
  components: {
  },
  data() {
    return {
      errorMessage: undefined,
      dialog: true,
      business: '',
      description: '',
      businessType: [],
      street1: '',
      district: '',
      city: '',
      region: '',
      country: '',
      postcode: '',
      businessTypes: [
        'Accommodation and Food Services',
        'Charitable organisation',
        'Non-profit organisation',
        'Retail Trade',
      ],
      valid: false,
      maxCharRules: [
        field => (field.length <= 100) || 'Reached max character limit: 100'
      ],
      maxCharDescriptionRules: [
        field => (field.length <= 200) || 'Reached max character limit: 200'
      ],
      mandatoryRules: [
        //All fields with the class "required" will go through this ruleset to ensure the field is not empty.
        //if it does not follow the format, display error message
        field => !!field || 'Field is required'
      ],
      streetRules: [
        field => /^(?=.*[0-9 ])(?=.*[\s])(?=.*[a-zA-Z ])([a-zA-Z0-9 ]+)$/.test(field) || 'Must have at least one number and one alphabet'
      ]
    };
  },
  methods: {
    async createBusiness() {
      this.errorMessage = undefined;
      /**
       * Get the street number and name from the street address field.
       */
      const streetParts = this.street1.split(" ");
      const streetNum = streetParts[0];
      const streetName = streetParts.slice(1, streetParts.length).join(" ");
      let business = {
        primaryAdministratorId: this.$store.state.user.id,
        name: this.business,
        description: this.description,
        address: {
          streetNumber: streetNum,
          streetName: streetName,
          district: this.district,
          city: this.city,
          region: this.region,
          country: this.country,
          postcode: this.postcode,
        },
        businessType: this.businessType,
      };
      let response = await createBusiness(business);
      if (response === undefined) {
        this.closeDialog();
        this.$router.go();
      } else {
        this.errorMessage = response;
      }
    },
    closeDialog() {
      this.$emit('closeDialog');
    }
  }
};
</script>

<style scoped>
/* Mandatory fields are accompanied with a * after it's respective labels*/
.required label::after {
  content: "*";
  color: red;
}
</style>