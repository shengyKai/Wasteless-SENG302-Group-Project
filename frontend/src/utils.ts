import { MarketplaceCardSection } from "./api/internal";

/**
 * Sets a cookie. Will expire in one year.
 * @param name Name of the cookie being set.
 * @param value Value to store in the cookie.
 */
export function setCookie(name: string, value: string|number) {
  const date = new Date();
  date.setFullYear(date.getFullYear() + 1);
  document.cookie = `${name}=${value}; expires=${date.toUTCString()}; path=/`;
}

/**
 * Reads a cookie.
 * @param name Name of the cookie to read.
 * @returns Value of the cookie. NULL if the cookie does not exist.
 */
export function getCookie(name: string) {
  let cookies = document.cookie.split(';');
  for (let cookie of cookies) {
    cookie = cookie.trim();
    if (cookie.startsWith(name + '=')) {
      return cookie;
    }
  }
  return null;
}

/**
 * Deletes a cookie.
 * @param name Name of the cookie to delete.
 */
export function deleteCookie(name: string) {
  document.cookie = `${name}=; Path=/; Expires=Thu, 01 Jan 1970 00:00:01 GMT;`;
}

/**
 * Cookie names.
 */
export const COOKIE = {
  USER: 'user'
};

/**
 * Converts a date to a human readable string
 *
 * @param date Date to format, either a date object or a parsable date string
 * @returns Date in a human readable string
 */
export function formatDate(date: Date | string) {
  if (typeof date === 'string') date = new Date(date);
  const parts = date.toDateString().split(' ');
  return `${parts[2]} ${parts[1]} ${parts[3]}`;
}

export function formatPrice(price : number) {
  if (Number.isInteger(price)) {
    return price.toString();
  } else {
    return (+price).toFixed(2);
  }
}

/**
 * User roles.
 */
export const USER_ROLES = {
  USER: 'user',
  GAA: 'globalApplicationAdmin',
  DGAA: 'defaultGlobalApplicationAdmin'
};

/**
 * Mapping between section key and section name
 */
export const SECTION_NAMES: Record<MarketplaceCardSection, string> = {
  ForSale: 'For Sale',
  Wanted: 'Wanted',
  Exchange: 'Exchange',
};


/**
 * Rate limits the input function by waiting a given amount of time calling the inner function.
 * If interrupted with another call then this wait is reset.
 *
 * @param func Function to rate limit
 * @param wait Time (ms) to wait before calling the function
 */
export function debounce(func: (() => void), wait: number) {
  let timeout: number | undefined;
  function debounced() {
    if (timeout !== undefined) {
      clearTimeout(timeout);
    }
    timeout = setTimeout(func, wait);
  }
  return debounced;
}

/**
 * Trims the input string such that it is always less than or equal to the provided length
 * This function will try to avoid breaking up words unless there is no alternative
 * @param str String to trim
 * @param length Maximum length
 * @returns Trimmed string
 */
export function trimToLength(str: string, length: number) {
  const re = new RegExp(`^(.{0,${length}})(\\s.*|$)`);
  return str.replace(re, "$1").slice(0, length);
}

/**
 * Checks whether we are executing in the test enviroment
 *
 * @returns true if we're testing false otherwise
 */
export function isTesting() {
  return process.env.JEST_WORKER_ID !== undefined;
}

/**
 * ========= REGEX =========
 */

/**
 * Returns a Regex pattern that matches letters, spaces, dashes, hyphen and apostrophe
 */
export function regxAlphabet() {
  return /^[\p{L}\-.' ]*$/u;
}

/**
 * Returns a Regex pattern that matches letters, spaces, numbers and common characters
 */
export function regxAlphabetExtended() {
  return /(^[\p{L}\d\p{P} ]*$)/u;
}

export function regxAlphabetExtendedMultiline() {
  return /(^[\p{L}\d\p{P}$\s]*$)/u;
}

/**
 * Returns a Regex pattern that matches numbers
 */
export function regxNumerical() {
  return /(^[\d]*$)/;
}

/**
 * Returns a Regex pattern that matches a price
 */
export function regxPrice() {
  return /(^\d{1,8}(\.\d{2})?$)|^$/;
}

/**
 * Returns a Regex that matches a password with at least one number and one letter.
 */
export function regxPassword() {
  return /^(?=.*[0-9])(?=.*[\p{L} ])([\p{L}0-9 !@#$%^&*()-]+)$/u;
}

/**
 * Returns a Regex that matches a valid email address
 */
export function regxEmail() {
  return /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
}

/**
 * Returns a Regex that matches a valid post code
 */
export function regxPostCode() {
  return /(^[\p{L}0-9]*$)/u;
}

/**
 Returns a Regex that matches a valid phone number
 */
export function regxPhoneNumber() {
  return /(^\(?\d{1,3}\)?[\s.-]?\d{3,4}[\s.-]?\d{4,5}$)|(^$)/;
}

/**
 Returns a Regex that matches a country code
 */
export function regxCountryCode() {
  return /(^(\d{1,2}-)?\d{2}$)|(^$)/;
}
/**
 Returns a Regex that matches a valid street address
 */
export function regxStreet() {
  return /^(([0-9]+|[0-9]+\/[0-9]+)[\p{L}]?)(?=\s[\p{L}])([\p{L}0-9-.' ]+)$/u;
}
/**
 Returns a Regex that matches a valid product code
 */
export function regxProductCode() {
  return /^[-A-Z0-9]+$/;
}

/**
 * ========= RULES =========
 */

export const emailRules = [
  (email: string) => regxEmail().test(email) || 'E-mail must be valid'
];

export const mandatoryRules = [
  //All fields with the class "required" will go through this ruleset to ensure the field is not empty.
  //if it does not follow the format, display error message
  (field: string) => !!field || 'Field is required'
];

export const passwordRules = [
  (field: string) => (field && field.length >= 7) || 'Password must have 7+ characters',
  (field: string) => regxPassword().test(field) || 'Must have at least one number and one letter'
];

export const postCodeRules = [
  (field: string) => regxPostCode().test(field) || 'Must contain numbers and alphabet only'
];

export const nameRules = [
  (field: string) =>  (field.length === 0 || regxAlphabet().test(field)) || 'Naming must only contain letter and common characters'
];

export function maxCharRules(size: number) {
  return [
    (field: string) => (field.length <= size) || `Reached max character limit: ${size}`
  ];
}

export const phoneNumberRules = [
  (field: string) => regxPhoneNumber().test(field) || 'Must only contain 8-12 digits'
];
export const countryCodeRules = [
  (field: string) => regxCountryCode().test(field) || 'Must only contain 2 digits.'
];
export const alphabetRules = [
  (field: string) => (field.length === 0 || regxAlphabet().test(field)) || 'Naming must be valid'
];
export const streetNumRules = [
  (field: string) => (field && field.length <= 109) || 'Reached max character limit 109 ',
  (field: string) => regxStreet().test(field) || 'Must contain unit number and street name'
];

export const quantityRules = [
  (field: string) => regxNumerical().test(field) || 'Must contain an integer',
  (field: string) => field ? parseInt(field) > 0 || 'Must be greater than zero' : true
];

export const smallPriceRules = [
  //A price must be numbers and may contain a decimal followed by exactly two numbers (4digit)
  (field: string) => regxPrice().test(field) || 'Must be a valid price',
  (field: string) => field ? parseInt(field) < 10000 || 'Must be less than 10,000' : true,
];

export const hugePriceRules = [
  //A price must be numbers and may contain a decimal followed by exactly two numbers (6digit)
  (field: string) => regxPrice().test(field) || 'Must be a valid price',
  (field: string) => field ? parseInt(field) < 1000000 || 'Must be less than 1,000,000' : true
];

export const alphabetExtendedSingleLineRules = [
  (field: string) => regxAlphabetExtended().test(field) || 'Must only contain letters, numbers, punctuation and spaces',
];

export const alphabetExtendedMultilineRules = [
  (field: string) => regxAlphabetExtendedMultiline().test(field) || 'Must only contain letters, numbers, punctuation and whitespace',
];

export const productCodeRules = [
  (field: string) => field.length <= 15 || 'Reached max character limit: 15',
  (field: string) => !/ /.test(field) || 'Must not contain a space',
  (field: string) => regxProductCode().test(field) || 'Must be all uppercase letters, numbers and dashes.',
];
