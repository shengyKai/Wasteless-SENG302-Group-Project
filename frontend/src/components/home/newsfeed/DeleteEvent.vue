<template>
  <div>
    <v-card-title>
      {{ title }}
    </v-card-title>
    <v-card-subtitle>
      {{ date }}
    </v-card-subtitle>
    <v-card-text>
      {{ text }}
    </v-card-text>
  </div>
</template>

<script>
import { formatDate } from '@/utils';

export default {
  name: 'DeleteEvent',
  props: {
    event: {
      type: Object
    },
  },
  data() {
    return {
      title: "Your marketplace card has been deleted"
    };
  },
  computed: {
    text() {
      return `Your marketplace card "${this.event.title}" in the ${this.section} section has been removed from the marketplace`;
    },
    section() {
      if (this.event.section === "ForSale") {
        return "For Sale";
      }
      return this.event.section;
    },
    date() {
      return formatDate(this.event.created);
    }
  }
};
</script>