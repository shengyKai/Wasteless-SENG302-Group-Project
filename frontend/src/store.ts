import {User,  Business, getUser, login} from './api';
import Vuex, { Store, StoreOptions } from 'vuex';
import { COOKIE, deleteCookie, setCookie } from './utils';

export type StoreData = {
  user: User | null,
  activeRole: { type: "user" | "business", id: number} | null,
};

export function createOptions(): StoreOptions<StoreData> {
  return {
    state: {
      user: null,
      activeRole: null,
    },
    mutations: {
      setUser (state, payload: User) {
        state.user = payload;
        // If the payload contains a user ID, user is now logged in. Set their session cookie.
        if (payload.id) {
          deleteCookie(COOKIE.USER.toUpperCase());
          deleteCookie(COOKIE.USER.toLowerCase());
          setCookie(COOKIE.USER, payload.id);
        }
      },
      logoutUser (state) {
        state.user = null;
        deleteCookie(COOKIE.USER);
      }
    },
    getters: {
      isLoggedIn (state) {
        return state.user !== null;
      },
      role (state) {
        return state.user?.role;
      }
    },
    actions: {
      async login(context, { email, password }) {
        let userId = await login(email, password);
        if (typeof userId === 'string') {
          console.warn(userId);
          return;
        }
        let user = await getUser(userId);
        if (typeof user === 'string') {
          console.error(user);
          return;
        }
        context.commit('setUser', user);
      }
    }
  };
}

export default function createStore() {
  return new Vuex.Store(createOptions());
}
