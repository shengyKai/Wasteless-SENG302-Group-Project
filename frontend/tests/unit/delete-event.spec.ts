
import Vue from 'vue';
import Vuetify from 'vuetify';
import { createLocalVue, mount, Wrapper } from '@vue/test-utils';
import DeleteEvent from '@/components/home/newsfeed/DeleteEvent.vue';

import Vuex, { Store } from 'vuex';
import { getStore, resetStoreForTesting, StoreData } from '@/store';
import { makeTestUser } from './utils';

Vue.use(Vuetify);

describe('DeleteEvent.vue', () => {
  let wrapper: Wrapper<any>;
  let vuetify: Vuetify;
  // The global store to be used
  let store: Store<StoreData>;

  beforeEach(async () => {
    const localVue = createLocalVue();
    vuetify = new Vuetify();

    localVue.use(Vuex);
    resetStoreForTesting();
    store = getStore();
    store.state.user = makeTestUser(1);

    const app = document.createElement ("div");
    app.setAttribute ("data-app", "true");
    document.body.append (app);

    wrapper = mount(DeleteEvent, {
      localVue,
      vuetify,
      store,
      propsData: {
        event: {
          id: 101,
          title: "Card title",
          section: "ForSale",
          created: "2021-01-01T12:00:00Z",
          type: "DeleteEvent"
        },
      }
    });
  });

  describe('Card has been deleted from the database', () => {
    it("Title contains delete message", () => {
      let eventWrapper = wrapper.findComponent({ name: 'Event'});
      expect(eventWrapper.props().title).toBe('Your marketplace card has been deleted');
    });

    it("Text contains delete message if the card is in the ForSale section ", () => {
      expect(wrapper.vm.text).toBe('Your marketplace card "Card title" in the For Sale section has been removed from the marketplace');
    });

    it("Text contains delete message if the card is in other sections other than ForSale", async () => {
      await wrapper.setProps({
        event: {
          id: 101,
          title: "Card title",
          section: "Wanted",
          created: "2021-01-01T12:00:00Z",
          type: "DeleteEvent"
        },
      });
      expect(wrapper.vm.text).toBe('Your marketplace card "Card title" in the Wanted section has been removed from the marketplace');
    });

    it('Must match snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});