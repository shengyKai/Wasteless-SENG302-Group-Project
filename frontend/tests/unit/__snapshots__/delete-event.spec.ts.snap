// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeleteEvent.vue Card has been deleted from the database Must match snapshot 1`] = `
<div>
  <div class="delete"><button type="button" class="v-icon notranslate deleteButton v-icon--link mdi mdi-trash-can theme--light red--text"></button>
    <div class="deletion-error">

    </div>
  </div>
  <div class="v-card__title">
    Your marketplace card has been deleted
  </div>
  <div class="v-card__subtitle">
    02 Jan 2021, 01:00
  </div>
  <div class="v-card__text">
    Your marketplace card "Card title" in the For Sale section has been removed from the marketplace
  </div>
</div>
`;
